File: android/src/main/kotlin/dev/doclens/doclens/ScannerConfig.kt
package dev.doclens.doclens

data class ScannerConfig(
    val enableLiveDetection: Boolean,
    val detectionThrottleHz: Int,
    val enablePerspectiveWarp: Boolean,
    val jpegQuality: Int,
    val outputFormat: String,
    val captureResolution: String,
    val initialFlashMode: String,
    val initialLens: String,
    val enableTapToFocus: Boolean,
    val enablePinchToZoom: Boolean,
    val enableLowLightDetection: Boolean,
) {
    companion object {
        fun fromMap(m: Map<String, Any?>): ScannerConfig = ScannerConfig(
            enableLiveDetection = m["enableLiveDetection"] as? Boolean ?: true,
            detectionThrottleHz = (m["detectionThrottleHz"] as? Number)?.toInt() ?: 12,
            enablePerspectiveWarp = m["enablePerspectiveWarp"] as? Boolean ?: true,
            jpegQuality = (m["jpegQuality"] as? Number)?.toInt() ?: 92,
            outputFormat = m["outputFormat"] as? String ?: "jpeg",
            captureResolution = m["captureResolution"] as? String ?: "high",
            initialFlashMode = m["initialFlashMode"] as? String ?: "auto",
            initialLens = m["initialLens"] as? String ?: "back",
            enableTapToFocus = m["enableTapToFocus"] as? Boolean ?: true,
            enablePinchToZoom = m["enablePinchToZoom"] as? Boolean ?: true,
            enableLowLightDetection = m["enableLowLightDetection"] as? Boolean ?: true,
        )
    }
}

data class Quad(
    val topLeft: PointF,
    val topRight: PointF,
    val bottomRight: PointF,
    val bottomLeft: PointF,
) {
    fun toMap(): Map<String, Any?> = mapOf(
        "topLeft" to listOf(topLeft.x.toDouble(), topLeft.y.toDouble()),
        "topRight" to listOf(topRight.x.toDouble(), topRight.y.toDouble()),
        "bottomRight" to listOf(bottomRight.x.toDouble(), bottomRight.y.toDouble()),
        "bottomLeft" to listOf(bottomLeft.x.toDouble(), bottomLeft.y.toDouble()),
    )

    companion object {
        @Suppress("UNCHECKED_CAST")
        fun fromMap(m: Map<String, Any?>): Quad {
            fun pt(key: String): PointF {
                val raw = m[key] as? List<Number> ?: return PointF(0f, 0f)
                return PointF(raw[0].toFloat(), raw[1].toFloat())
            }
            return Quad(pt("topLeft"), pt("topRight"), pt("bottomRight"), pt("bottomLeft"))
        }
    }
}

data class PointF(val x: Float, val y: Float)

sealed class ScannerException(message: String) : Exception(message) {
    class PermissionDenied : ScannerException("Camera permission denied")
    class Unavailable(message: String) : ScannerException(message)
    class InitFailed(message: String) : ScannerException(message)
    class CaptureFailed(message: String) : ScannerException(message)
}

File: ios/Classes/ScannerConfig.swift
import Foundation
import AVFoundation

struct ScannerConfig {
    let enableLiveDetection: Bool
    let detectionThrottleHz: Int
    let enablePerspectiveWarp: Bool
    let jpegQuality: Int
    let imageEnhancement: String
    let outputFormat: String
    let captureResolution: String
    let initialFlashMode: String
    let initialLens: String
    let enableTapToFocus: Bool
    let enablePinchToZoom: Bool
    let enableLowLightDetection: Bool

    static func from(map: [String: Any]) -> ScannerConfig {
        return ScannerConfig(
            enableLiveDetection: map["enableLiveDetection"] as? Bool ?? true,
            detectionThrottleHz: map["detectionThrottleHz"] as? Int ?? 12,
            enablePerspectiveWarp: map["enablePerspectiveWarp"] as? Bool ?? true,
            jpegQuality: map["jpegQuality"] as? Int ?? 100,
            imageEnhancement: map["imageEnhancement"] as? String ?? "none",
            outputFormat: map["outputFormat"] as? String ?? "jpeg",
            captureResolution: map["captureResolution"] as? String ?? "high",
            initialFlashMode: map["initialFlashMode"] as? String ?? "auto",
            initialLens: map["initialLens"] as? String ?? "back",
            enableTapToFocus: map["enableTapToFocus"] as? Bool ?? true,
            enablePinchToZoom: map["enablePinchToZoom"] as? Bool ?? true,
            enableLowLightDetection: map["enableLowLightDetection"] as? Bool ?? true
        )
    }

    var sessionPreset: AVCaptureSession.Preset {
        switch captureResolution {
        case "max": return .photo
        case "high": return .high
        default: return .high
        }
    }
}

struct Quad {
    let topLeft: CGPoint
    let topRight: CGPoint
    let bottomRight: CGPoint
    let bottomLeft: CGPoint

    func toMap() -> [String: Any] {
        return [
            "topLeft": [Double(topLeft.x), Double(topLeft.y)],
            "topRight": [Double(topRight.x), Double(topRight.y)],
            "bottomRight": [Double(bottomRight.x), Double(bottomRight.y)],
            "bottomLeft": [Double(bottomLeft.x), Double(bottomLeft.y)],
        ]
    }

    static func fromMap(_ map: [String: Any]) -> Quad {
        func pt(_ key: String) -> CGPoint {
            if let arr = map[key] as? [Double], arr.count == 2 {
                return CGPoint(x: arr[0], y: arr[1])
            }
            if let arr = map[key] as? [NSNumber], arr.count == 2 {
                return CGPoint(x: arr[0].doubleValue, y: arr[1].doubleValue)
            }
            return .zero
        }
        return Quad(topLeft: pt("topLeft"),
                    topRight: pt("topRight"),
                    bottomRight: pt("bottomRight"),
                    bottomLeft: pt("bottomLeft"))
    }
}

enum ScannerError: Error, LocalizedError {
    case permissionDenied
    case unavailable(String)
    case initFailed(String)
    case captureFailed(String)

    var errorDescription: String? {
        switch self {
        case .permissionDenied: return "Camera permission denied"
        case .unavailable(let m): return m
        case .initFailed(let m): return m
        case .captureFailed(let m): return m
        }
    }
}
